core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -154,14 +160,25 @@ private class ShuffleStatus(
    */
   def updateMapOutput(mapId: Long, bmAddress: BlockManagerId): Unit = withWriteLock {
     try {
-      val mapStatusOpt = mapStatuses.find(_.mapId == mapId)
+      val mapStatusOpt = mapStatuses.find(x => x != null && x.mapId == mapId)
       mapStatusOpt match {
         case Some(mapStatus) =>
           logInfo(s"Updating map output for ${mapId} to ${bmAddress}")
           mapStatus.updateLocation(bmAddress)
           invalidateSerializedMapOutputStatusCache()
         case None =>
-          logWarning(s"Asked to update map output ${mapId} for untracked map status.")
+          val index = mapStatusesDeleted.indexWhere(x => x != null && x.mapId == mapId)
+          if (index >= 0) {
+            val mapStatus = mapStatusesDeleted(index)
+            mapStatus.updateLocation(bmAddress)
+            assert(mapStatuses(index) == null)
+            mapStatuses(index) = mapStatus
+            _numAvailableMapOutputs += 1
+            invalidateSerializedMapOutputStatusCache()
+            logDebug(s"Recover ${mapStatus.mapId} ${mapStatus.location}")
```

Comment:
Do we need to clear `mapStatusesDeleted(index)`?

Reply:
No, we should not clear `mapStatusesDeleted` after this recovery. We may need to recover more than once if this executor dies again.

Reply:
But that will trigger another `removeMapOutput` right ?
And so get saved in `mapStatusesDeleted` at that time ?

Reply:
Right, it will be matched at that time. I'll update the PR.